src/res/heatmap.html

```diff
@@ -0,0 +1,4294 @@
+				}
+			}
+		}
+		c.strokeRect(0, 7 * 16, w, 16);
+	}
+
+	canvas.onmousemove = function (e) {
+		previewX = e.offsetX;
+		previewY = e.offsetY;
+		if (e.shiftKey) {
+			showPreview();
+		} else {
+			preview_wrapper.style.display = 'none';
+		}
+
+		const h = Math.floor(e.offsetY / 16);
+		if (h >= 0 && h < levels.length) {
+			const f = findFrame(levels[h], e.offsetX / px + root.left);
+			if (f) {
+				hl.style.left = (Math.max(f.left - root.left, 0) * px + canvas.offsetLeft) + 'px';
+				hl.style.width = (Math.min(f.width, root.width) * px) + 'px';
+				hl.style.top = (h * 16 + canvas.offsetTop) + 'px';
+				hl.firstChild.textContent = f.title;
+				hl.style.display = 'block';
+				let details = f.details || '';
+				canvas.title = f.title + '\n(' + samples(f.width) + details + ', ' + pct(f.width, levels[0][0].width) + '%)';
+				canvas.style.cursor = 'pointer';
+				canvas.onclick = function (e) {
+					if (e.shiftKey) {
+						e.preventDefault();
+						searchForMethod(f.method);
+						canvas.onmousemove(e);
+						return;
+					}
+
+					if (e.ctrlKey) {
+						const n = e.shiftKey ? findPrevFame(levels[h], e.offsetX / px + root.left) : findNextFrame(levels[h], e.offsetX / px + root.left);
+						if (n) {
+							render(n, h);
+						}
+						canvas.onmousemove(e);
+						return;
+					}
+
+					if (f !== root) {
+						render(f, h);
+						canvas.onmousemove(e);
+					}
+				};
+				status.textContent = 'Function: ' + canvas.title;
+				status.style.display = 'inline';
+				return;
+			}
+		}
+		canvas.onmouseout();
+	}
+
+	canvas.onmouseout = function (e) {
+		hl.style.display = 'none';
+		status.textContent = '\xa0';
+		status.style.display = 'none';
+		canvas.title = '';
+		canvas.style.cursor = '';
+		canvas.onclick = '';
+		if (e !== undefined) {
+			preview_wrapper.style.display = 'none';
+			previewX = previewY = -1;
+		}
+	}
+
+	document.getElementById('reset').onclick = function () {
+		search(false);
+	}
+
+	class DataBuffer {
+		data;
+		pos = 0;
+
+		constructor(encodedData) {
+			this.data = encodedData;
+		}
+
+		nextVarInt() {
+			let res = 0;
+			let shift = 1;
+			let b;
+			do {
+				b = this.byteAt(this.pos++);
+				res += (b % 61) * shift;
+				shift *= 61;
+			} while (b >= 61);
+			return res;
+		}
+
+		nextInt6() {
+			return this.byteAt(this.pos++);
+		}
+
+		nextSymbol() {
+			return this.byteAt(this.pos++);
+		}
+
+		nextInt30() {
+			return this.byteAt(this.pos++) << 0
+				| this.byteAt(this.pos++) << 6
+				| this.byteAt(this.pos++) << 12
+				| this.byteAt(this.pos++) << 18
+				| this.byteAt(this.pos++) << 24;
+		}
+
+		byteAt(pos) {
+			const c = this.data.charCodeAt(pos);
+			switch (c) {
+				case 127:
+					return 0;
+				case 126:
+					return 13;
+				case 125:
+					return 38;
+				case 124:
+					return 60;
+				case 123:
+					return 62;
+			}
+			return c;
+		}
+
+		nextInt18() {
+			return this.nextInt6() | this.nextInt6() << 6 | this.nextInt6() << 12;
+		}
+
+		int30(pos) {
```

Comment:
Instead of this function, I'd just set `pos` to the tail of the stream and reuse `nextInt30`.

Reply:
I don't think touching pos is a good idea, nice place for bugs.